components/raftstore-v2/src/operation/command/admin/split.rs

```diff
@@ -0,0 +1,569 @@
+
+        self.region_state_mut().set_region(derived.clone());
+        self.region_state_mut().set_tablet_index(log_index);
+        // self.metrics.size_diff_hint = 0;
+        // self.metrics.delete_keys_hint = 0;
+
+        let mut resp = AdminResponse::default();
+        resp.mut_splits().set_regions(regions.clone().into());
+        PEER_ADMIN_CMD_COUNTER.batch_split.success.inc();
+
+        Ok((
+            resp,
+            AdminCmdResult::SplitRegion(SplitResult {
+                regions,
+                derived,
+                tablet_index: log_index,
+            }),
+        ))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::sync::{
+        mpsc::{channel, Receiver, Sender},
+        Arc,
+    };
+
+    use collections::HashMap;
+    use engine_test::{
+        ctor::{CfOptions, DbOptions},
+        kv::TestTabletFactoryV2,
+        raft,
+    };
+    use engine_traits::{CfOptionsExt, Peekable, WriteBatch, ALL_CFS};
+    use futures::channel::mpsc::unbounded;
+    use kvproto::{
+        raft_cmdpb::{AdminCmdType, BatchSplitRequest, PutRequest, RaftCmdResponse, SplitRequest},
+        raft_serverpb::{PeerState, RaftApplyState, RegionLocalState},
+    };
+    use raftstore::store::{cmd_resp::new_error, Config, ReadRunner};
+    use slog::o;
+    use tempfile::TempDir;
+    use tikv_util::{
+        codec::bytes::encode_bytes,
+        config::VersionTrack,
+        store::{new_learner_peer, new_peer},
+        worker::{dummy_future_scheduler, dummy_scheduler, FutureScheduler, Scheduler, Worker},
+    };
+
+    use super::*;
+    use crate::{
+        fsm::{ApplyFsm, ApplyResReporter},
+        raft::Apply,
+        tablet::CachedTablet,
+    };
+
+    struct MockReporter {
+        sender: Sender<ApplyRes>,
+    }
+
+    impl MockReporter {
+        fn new() -> (Self, Receiver<ApplyRes>) {
+            let (tx, rx) = channel();
+            (MockReporter { sender: tx }, rx)
+        }
+    }
+
+    impl ApplyResReporter for MockReporter {
+        fn report(&self, apply_res: ApplyRes) {
+            let _ = self.sender.send(apply_res);
+        }
+    }
+
+    fn new_split_req(key: &[u8], id: u64, children: Vec<u64>) -> SplitRequest {
+        let mut req = SplitRequest::default();
+        req.set_split_key(key.to_vec());
+        req.set_new_region_id(id);
+        req.set_new_peer_ids(children);
+        req
+    }
+
+    fn assert_split(
+        apply: &mut Apply<engine_test::kv::KvTestEngine, MockReporter>,
+        factory: &Arc<TestTabletFactoryV2>,
+        region_to_split: &Region,
+        right_derived: bool,
+        new_region_ids: Vec<u64>,
+        split_keys: Vec<Vec<u8>>,
+        children_peers: Vec<Vec<u64>>,
+        log_index: u64,
+    ) -> HashMap<u64, Region> {
+        let mut splits = BatchSplitRequest::default();
+        splits.set_right_derive(right_derived);
+        let mut region_boundries = split_keys.clone();
+        region_boundries.insert(0, region_to_split.get_start_key().to_vec());
```

Comment:
Pointless to repeat what has done in apply. You should use table based test cases and specify boundaries explicitly.